### PR TITLE
Add 'add using' support in Razor.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/EditorFeatures/Core.Cocoa/Snippets/AbstractSnippetExpansionClient.cs [hunk 1]
```diff
@@ -51,7 +51,7 @@ public AbstractSnippetExpansionClient(IThreadingContext threadingContext, IConte
 
         public abstract IExpansionFunction GetExpansionFunction(XElement xmlFunctionNode, string fieldName);
         protected abstract ITrackingSpan InsertEmptyCommentAndGetEndPositionTrackingSpan();
-        internal abstract Document AddImports(Document document, int position, XElement snippetNode, bool placeSystemNamespaceFirst, CancellationToken cancellationToken);
+        internal abstract Document AddImports(Document document, int position, XElement snippetNode, bool placeSystemNamespaceFirst, bool allowInHiddenRegions, CancellationToken cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Cocoa/Snippets/AbstractSnippetExpansionClient.cs:54]: the majority of this PR is just threading this option through *everything.........*

**Comment by TanayParikh** [comment]
@CyrusNajmabadi is this ready to try out?

**Comment by CyrusNajmabadi** [comment]
@TanayParikh Yes it is.

**Review by davidwengier — APPROVED** [review]

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/CompilationUnitSyntaxExtensions.cs [hunk 2]
```diff
@@ -16,16 +16,19 @@ namespace Microsoft.CodeAnalysis.CSharp.Extensions
...
             var usingDirectiveAncestor = contextNode.GetAncestor<UsingDirectiveSyntax>();
-            if ((usingDirectiveAncestor != null) && (usingDirectiveAncestor.GetAncestor<NamespaceDeclarationSyntax>() == null))
+            if (usingDirectiveAncestor?.Parent is CompilationUnitSyntax)
             {
                 // We are inside a top level using directive (i.e. one that's directly in the compilation unit).
                 return false;
             }
 
-            if (contextNode.SyntaxTree.HasHiddenRegions())
+            if (!allowInHiddenRegions && contextNode.SyntaxTree.HasHiddenRegions())
```
Comment (davidwengier) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/CompilationUnitSyntaxExtensions.cs:31]: Probably just me, but for some reason I had to stare at this for a few minutes before it made sense, but I got there in there end. Something about the "not allow", versus "has hidden" ¯\\\_(ツ)_/¯

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/CompilationUnitSyntaxExtensions.cs:31]: yeah.  my brain has the same issue at times.  If this were in a lot of areas, i'd def want to simplify.  But i don't mind it right here.   :)

**Pull request merged**